### DISCUSSION TITLE
v3.0.0-rc.13

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#16409] [Announcements]
> **Note**
> We are in the process of finalizing Nuxt 3. 1 or 2 additional release candidates are expected before the final 3.0.0 version.

👉  **[Full Changelog](https://github.com/nuxt/framework/releases/tag/v3.0.0-rc.13)**

### 🔰 Security Fixes

This release contains multiple security related fixes [#8675](https://github.com/nuxt/framework/pull/8675), [#8674](https://github.com/nuxt/framework/pull/8674) and [#8673](https://github.com/nuxt/framework/pull/8673) reported via [huntr.dev](https://huntr.dev/) platform by [OhB00](https://github.com/OhB00).

We recommend you upgrade to the latest version as soon as possible.

If you encounter "The request URL ... is outside of Vite service allow list" issue, try adding path to `vite.server.fs.allow` in `nuxt.config`. [read more](https://github.com/nuxt/framework/pull/8674).

### 🚀 Performance Improvements

Using a new method to extract `definePageMeta` improves vite performance and makes lazy compilation of pages possible ([#8536](https://github.com/nuxt/framework/pull/8536)).

### 💯 Strict Config Schema and Types

We have cleaned up the configuration schema ([#8487](https://github.com/nuxt/framework/pull/8487)) so that you no longer would be confused with Nuxt 2 options and also can quickly notice any typos in `nuxt.config` file.

Typescript strict mode is also enabled by default with this release as best practice. ([#8667](https://github.com/nuxt/framework/pull/8667))

### 🚇 Nitro Development Server Proxy

Using `nitro.devProxy` option you can now configure proxies for the development server. ([learn more](https://nitro.unjs.io/config#devproxy))

We are working to support production runtime proxy in future Nitro versions using route rules.

---

👉  **[Full Changelog](https://github.com/nuxt/framework/releases/tag/v3.0.0-rc.13)**

**Comment by FileSoft-us** [comment]
Thanks you all

**Comment by tao** [comment]
I just updated but it seems tailwindcss stopped working, any ideas why? Is it the change to `nuxt.config` file?

```js
export default defineNuxtConfig({
  css: ['~/assets/css/tailwind.css'],
  watch: ['~/postcss.config.js', '~/tailwind.config.js'],
  build: {
    postcss: {
      postcssOptions: require('./postcss.config.js'),
    },
  }
})
```

**Reply by tao** [reply]
Ok, so to fix it you can do this instead... and remove the `build` parent? Is that the correct approach?

```js
export default defineNuxtConfig({
  css: ['~/assets/css/tailwind.css'],
  watch: ['~/postcss.config.js', '~/tailwind.config.js'],
  postcss: require('./postcss.config.js'),
})
```

**Reply by calebsmithdev** [reply]
I use @nuxtjs/tailwincss and it's broken there as well, so I'll have to wait for them to update before I can continue with RC13.

**Reply by twuncher** [reply]
thanks that fix worked for me as well!

**Reply by danielroe** [reply]
The fix is already present in the latest version of the tailwind module.

**Reply by KareemDa** [reply]
@danielroe 
Yup, Updating `@nuxtjs/tailwindcss` fixed the issue. Thank you

**Reply by sjonibek** [reply]
Please reinstall Tailwindcss, everything will be OK! ;-)

**Reply by DASH0070** [reply]
> The fix is already present in the latest version of the tailwind module.

Can you link the solutions? Thank you

**Reply by FayKn** [reply]
> I use @nuxtjs/tailwincss and it's broken there as well, so I'll have to wait for them to update before I can continue with RC13.

Strange, for me it works on the newest version of tailwind (`npm uninstall @nuxtjs/tailwindcss`, then reinstall)

this is the config.ts part
```
    buildModules: [
        ["@nuxtjs/tailwindcss", {
            cssPath: "~/assets/css/tailwind.css",
            configPath: "tailwind.config.js",
            exposeConfig: false,
            config: {},
            injectPosition: 0,
            viewer: true
        }
        ],
    ],

    css: ["~/assets/css/tailwind.css"],
```
    
    
    and that works

**Comment by calebsmithdev** [comment]
Excited for the `<devOnly>` component! It's a small thing, but I can see it being very useful for some trial features I've been working on.

**Comment by slav-weber** [comment]
Nuxt is wonderful, want to have an update but after update I have always the same error when use `yarn install`
`ERROR  Cannot create property 'optimizeDeps' on boolean 'false'`
already removed node modules and clean nuxi cache, nothing helps
It looks like Nuxt trying to use Vite in a preparation process, but I dont want use Vite, already have it in my `nuxt.config.ts`:
```
vite: false, // I want to use webpack
typescript: { // Added this only when update to RC13, I think for me strict mode is not needed now
     strict: false
},
```
Please tell me if I need to open a ticket or if you need any more information

**Reply by danielroe** [reply]
If you want to use webpack you should set the `builder` option. False is not a valid value for `vite`.

(This was changed a number of releases ago.)

**Reply by slav-weber** [reply]
I did exactly as you said and it works. Thanks a lot.
I trying to find what you wrote about `builder` option in the documentations on both websites(links next) but still didn't found it, in the old documentation this information was on the page about 'nuxt.config.ts'
https://v3.nuxtjs.org/api/configuration/nuxt-config
https://nuxtjs.org/docs/directory-structure/nuxt-config
May be I'm blind, but please add this information to the documentation if it's not there and thanks again!)

**Reply by danielroe** [reply]
This is the section: https://v3.nuxtjs.org/api/configuration/nuxt-config#builder

I agree it should have information on how to switch.

**Reply by slav-weber** [reply]
ah, now I understand what the problem is, I often use ctrl + f to quickly find the content on the page and in the search for webpack I could not find the right one. I also thought "How strange that you can find anything on setting up webpack, but you can't find what activates it", everything turned out to be much simpler.

**Comment by dosstx** [comment]
Example of how to use this enhancement?
"nuxt: Make useFetch options reactive (https://github.com/nuxt/framework/pull/8374)"

**Reply by danielroe** [reply]
Here's an example: https://github.com/nuxt/framework/issues/8692#issuecomment-1303766010

**Comment by matteomejia** [comment]
I just upgraded to RC13 from RC12 and now my calls to useFetch return `<DataT | null>` instead of `<DataT>`. Is this intended behavior?

**Reply by danielroe** [reply]
Yes; we enabled strict mode by default; you can set `typescript.strict` to false to disable. I do think the async data type can be improved however.

**Reply by matteomejia** [reply]
Ok, I understand the reason now. Maybe I could split the AsyncData type into 2 and use a third type AsyncDataOrLazyAsyncData that changes the return type of the data depending on the lazy attribute? It can be done with the never type.

It sounds a bit too specific though, but I'd be happy to do it.

**Comment by DrJume** [comment]
Currently, it seems that on prerender every route is rendered completely new. If the Nuxt app has a large initial loading step, it is reflected in every route prerender time. When accessing the app through the browser, the initial load is only present on the first visit. All following page visits happen incrementally. 

Is it possible for the prerender process to only run the initial load once and "prerender" each page incrementally?

**Comment by Mensix** [comment]
I wouln't have came up with idea of `dev-only` component but now I see I definitely needed it 🤩

**Comment by safejace** [comment]
ran into a possible regression with `NuxtLayout` see discord thread for more details: https://discord.com/channels/473401852243869706/897487139888062506/1038203159610409111

getting 

``` 
[Vue warn]: Extraneous non-props attributes (page-data) were passed to component but could not be automatically inherited because component renders fragment or text root nodes.
```

**Reply by danielroe** [reply]
I am not sure why it was working before but it was never intended that you pass data to a page this way. The props passed to NuxtPage are not passed to your page; they are passed to the NuxtPage component.

**Comment by zhaoweizhong** [comment]
Good job! Are there any updates on the Auth module?

**Reply by kilakewe** [reply]
I'm eagerly waiting on this one too. It's the only thing holding us back from switching to Nuxt3.

**Comment by Tuguldurjs** [comment]
I think the nuxt/auth feature should be prioritized on top of those enhancements. Users will first enter the login/registration page.

**Comment by murugappanrm** [comment]
@tao 

I got the normal version of tailwindcss working with Nuxt3.  This is what i did:

(1) I installed tailwind using
```
npm install -D tailwindcss postcss autoprefixer
npx tailwind init
```
(2) My tailwind.config.js looks like this:

```
module.exports = {
  content: [
    "./components/**/*.{js,vue,ts}",
    "./layouts/**/*.vue",
    "./pages/**/*.vue",
    "./plugins/**/*.{js,ts}",
    "./nuxt.config.{js,ts}",
    "./**/*.vue",
  ],
  theme: {
    extend: {
      },
  },
  plugins: [],
};
```

(3) My nuxt.config.ts looks like this (without the build):
```

export default defineNuxtConfig({
  postcss: {
    plugins: {
      tailwindcss: {},
      autoprefixer: {},
    },
  },
  css: ["@/assets/css/tailwind.css", "@/assets/css/main.css"],
});

```

(4) The package.json looks like this

```
{
  "private": true,
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "preview": "nuxt preview",
    "postinstall": "nuxt prepare"
  },
  "devDependencies": {
    "autoprefixer": "^10.4.13",
    "nuxt": "3.0.0-rc.13",
    "postcss": "^8.4.18",
    "tailwindcss": "^3.2.2"
  }
}

```
This worked fine and no issue at all. I did not use the npm "@nuxtjs/tailwindcss" to avoid any issue with Tailwind upgrades.

**Reply by TheAlexLichter** [reply]
It is strongly suggested to use the tailwind module instead though. There won't be issues with minor tailwind version upgrades for this too.

**Comment by KareemDa** [comment]
I'm getting this error after upgrading: 
> No response returned from render handler: /__nuxt_error?url=/&statusCode=500&statusMessage=Vite+Error&message=[vite-node]+[ERR_INVALID_URL_SCHEME]+/@fs/D:/projects/ST+Mum/frontend/node_modules/.pnpm/ufo@0.8.6/node_modules/ufo/dist/index.mjs&stack=%3Cpre%3E%3Cspan+class=%22stack+internal%22%3Eat+/@fs/D:/projects/ST+Mum/frontend/node_modules/.pnpm/ufo@0.8.6/node_modules/ufo/dist/index.mjs%3C/span%3E%3C/pre%3E&data

**Reply by murugappanrm** [reply]
@KareemDa 

Not sure how to help. How and what was upgraded?

**Reply by TheAlexLichter** [reply]
Tracking in https://github.com/nuxt/framework/issues/8706

**Comment by mikesiiro** [comment]
You can't pass property devProxy to object NuxtConfig. 
So where do you add this if not in nuxt.config.ts ?

**Reply by TheAlexLichter** [reply]
You can add that via `nuxt.config.ts` in the `nitro` object. See e.g. https://github.com/albionstatus/albionstatus-website/blob/main/nuxt.config.ts#L24-L30

```ts
nitro: {
    devProxy: {
      '/api/': {
        target: 'https://your-endpoint.com/',
        prependPath: true,
        changeOrigin: true,
      },
    },
  },
  ```

**Reply by mikesiiro** [reply]
Thank you so much @manniL , it works!

**Comment by nestle49** [comment]
@pi0 
Could you confirm a release date, please?

First, the goal was June, then the middle of summer, then summer, then until September 23, then a few weeks next, then autumn, and now I see that in the last two PR it says that there will be 1-2 more PR - time stopped

We all understand that it's hard work. But we still would like to know at least an approximate real release date - december, january, february, spring? This is necessary in order to plan the migration

**Reply by mittci** [reply]
migration :(

**Reply by bhaskarGyan** [reply]
Wait is over.
16th November is the Nuxt v3 final release 

https://github.com/nuxt/framework/pull/8766/files

**Reply by mittci** [reply]
i migrate my nuxt 2 app some month ago, but because of bugs I abandoned it. 
now i have some warnings like:
- [Vue warn]: onServerPrefetch is called when there is no active component instance to be associated with. Lifecycle injection APIs can only be used during execution of setup(). If you are using async setup(), make sure to register lifecycle hooks before the first await statement.
- [nitro] [dev] [unhandledRejection] Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client at new NodeError (node:internal/errors:387:5)

**Reply by nestle49** [reply]
> https://github.com/nuxt/framework/pull/8766/files

It's not only about the release of the major version, but also the preparation of the intermediate - nuxt2, nuxt/bridge

**Reply by mittci** [reply]
And modules like auth (alternative @nuxtjs-alt/auth), http ($fetch and alternative @nuxtjs-alt/http), proxy, i18n (its ok).

**Comment by murugappanrm** [comment]
@bhaskarGyan 

The main problem is not the release of the software. Its the documentation that is equally important. In the absence of complete and working documentation, its does not matter what is released. The whole software will be anyone's guesswork and render it useless.

**Comment by Maximus1000** [comment]
Noting for others, on a minimal install of RC.13, I had an oh-my-fetch 500 API fetch error on attempting to get SSL/HTTPS configured with an mkcert certificate.  

The fix was as Nuxt core team has previously described ``NODE_TLS_REJECT_UNAUTHORIZED=0``.  I was able to replicate the 500 error in Node v18 easily then to fix with the above and obtain SSL/HTTPS in browser.  Nothing worked other than the above fix (I did not try downgrading Node).

I noted my problem [here (oh-my-fetch issue 156) ](https://github.com/unjs/ohmyfetch/issues/156#issuecomment-1310780502
) as I thought that it was related to oh-my-fetch and possibly Node v18.  If anyone has this issue with mkcert, see working solution there with nuxt.config and the dev script.

**Comment by murugappanrm** [comment]
Hi,

Suddenly the Nuxt3 install is not working anymore. I have done a few instances a few hours ago and they all work fine. Now any new installation is not working anymore. After i execute "npm run dev" i get the "Nitro" complete state. When click the URL  http://localhost:3000/ the page load up but i dont get the nuxt welcome  instead i get this:

`No response returned from render handler: /__nuxt_error?url=/&statusCode=500&statusMessage&message=head.addEntry+is+not+a+function&stack=%3Cpre%3E%3Cspan+class=%22stack+internal%22%3Eat+C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/nuxt/dist/head/runtime/lib/vueuse-head.plugin.mjs:11:8%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+fn+(C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/nuxt/dist/app/nuxt.mjs:148:27)%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+Object.callAsync+(/C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/unctx/dist/index.mjs:49:19)%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+callWithNuxt+(C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/nuxt/dist/app/nuxt.mjs:150:23)%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+applyPlugin+(C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/nuxt/dist/app/nuxt.mjs:94:29)%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+Module.applyPlugins+(C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/nuxt/dist/app/nuxt.mjs:104:11)%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+processTicksAndRejections+(node:internal/process/task_queues:96:5)%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+async+createNuxtAppServer+(C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/nuxt/dist/app/entry.mjs:30:7)%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+async+default+(/C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/@nuxt/vite-builder/dist/runtime/vite-node.mjs:27:18)%3C/span%3E%0A%3Cspan+class=%22stack+internal%22%3Eat+async+Object.renderToString+(/C:/xampp/htdocs/MyCoursesLG/nuxt3/nuxt3-bootstrap5cdn/nuxt-bscdn/node_modules/vue-bundle-renderer/dist/runtime.mjs:172:19)%3C/span%3E%3C/pre%3E&data`

Can anyone help?

**Reply by mittci** [reply]
- try to put it to a short path, without diacritics, uppercase letters, remove nuxt (nuxt 3) from path (like `c:\test`)
- run `npm run nuxi clean` and after `npm run dev`

**Comment by FayKn** [comment]
I see it's in the XAMP folder and maybe something is interfering there,
have you tried making it in a folder like documents?

**Comment by murugappanrm** [comment]
hi @Maxp777 and @Ovab 

I installed nuxt in **c:\test** folder with project name **nuxt-test**. Initial **npm run dev failed** the same as before. Then, as per @Maxp777 advise, i **did a "npx nuxi clean (not npm run nuxi clean)" and run "npm run dev". Failed as before**. To eliminate any virus interference, I did the same in another PC which has never run vue related products and i got the same failure. **I feel nuxt rc 13 is broken.**

**Reply by murugappanrm** [reply]
@Maxp777 looks like rc13 is broken. Same issues have been reported under issues and not fixed or replied!

**Comment by FileSoft-us** [comment]
__vite_ssr_import_0__.useNuxtApp(...)._useHead is not a function

**Comment by murugappanrm** [comment]
This is odd. Just a minute ago, I did a fresh Nuxt3 install and it seems to work fine. I did not do anything, just did a fresh install, thats it. The only thing i did different was to wait a full 1 minute after the "Nitro built" message before i loaded the URL "http://localhost:3000/

My theory in my case is that there is some background jobs running in my PC and Nitro build ma not be have completed yet even tho the message of completion is shown. I may be wrong. I am no expert in vuejs or nuxtjs. Just a rookie.
